### PR TITLE
Add piecewise_interpolate_schedule, linear_onecycle, and cos_onecycle.

### LINKED ISSUE #19: Implement OneCycle Learning Rate Schedule
OneCycle learning rate from this paper: https://arxiv.org/abs/1708.07120. 

The paper describes a 3-stage piecewise linear schedule. [PyTorch](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.OneCycleLR) and [FastAI](https://fastai1.fast.ai/callbacks.one_cycle.html) implement a 2 stage variant with cosine annealing (which apparently works better in practice).

I would like to contribute both to this library, if you think they are suitable additions.

### PR DESCRIPTION
**Pull request by n2cholas**
Closes #19.

* The parameters/docstrings for `cos_onecycle` and `linear_onecycle` are based on the [PyTorch version](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.OneCycleLR). 
* I'm not sure whether to name the methods `*_onecycle` or `*_onecycle_learning_rate`. The referenced paper also proposes a onecycle momentum schedule (implemented by PyTorch as well). If `optax` implements onecycle momentum in the future, it may be confusing.
* I tried to keep the API for `piecewise_interpolate_schedule` as close to `piecewise_constant_schedule` as I could. Please let me know if the arguments should be re-ordered, renamed, etc. In fact, now the `piecewise_constant_schedule` is just a special case of `piecewise_interpolate_schedule` (with a constant interpolation).
* What code format/linter does this library use? I just used [Google's pylintrc](https://google.github.io/styleguide/pyguide.html) and formatted manually, would be nice to include a `setup.cfg` or `.style.yapf` to auto-format code.

### CODE REVIEW

**Comment by mtthss** [comment]
Hi @n2cholas !
Apologies for the delay and thanks for the great contribution!

> The parameters/docstrings for cos_onecycle and linear_onecycle are based on the PyTorch version.

These look good!

> I'm not sure whether to name the methods

Current names seem fine

**Comment by mtthss** [comment]
> I tried to keep the API for piecewise_interpolate_schedule as close to piecewise_constant_schedule as I could

perfect!

File: optax/_src/schedule.py [hunk 1]
```diff
@@ -203,3 +203,137 @@ def schedule(count):
...
+  interval_sizes = (bounds[1:] - bounds[:-1])
+
+  def schedule(count):
+    indicator = (bounds[:-1] <= count) & (count < bounds[1:])
+    pct = (count - bounds[:-1]) / interval_sizes
+    interp_vals = interpolate_fn(values[:-1], values[1:], pct)
+    return indicator.dot(interp_vals) + (bounds[-1] <= count) * values[-1]
+
+  return schedule
+
```
Comment (mtthss) [inline optax/_src/schedule.py:263]: Need 2 blank lines before top level def.

File: optax/_src/schedule.py [hunk 2]
```diff
@@ -203,3 +203,137 @@ def schedule(count):
...
+  if transition_steps <= 0:
+    raise ValueError(
+        'A linear onecycle schedule was set with a non-positive '
+        '`transition_steps`')
+
+  return piecewise_interpolate_schedule(
+    'cosine',
+    peak_value / div_factor,
+    {int(pct_start * transition_steps): div_factor,
+     int(transition_steps): 1. / (div_factor * final_div_factor)})
```
Comment (mtthss) [inline optax/_src/schedule.py:339]: Wrong hanging indentation: Add 2 spaces

File: optax/_src/schedule_test.py [hunk 3]
```diff
@@ -302,5 +302,90 @@ def test_decay_count_greater_count_with_alpha(self):
         np.array(generated_vals), atol=1e-3)
 
 
+class PiecewiseInterpolateTest(chex.TestCase):
+
+  @chex.all_variants()
+  def test_linear_piecewise(self):
+    schedule_fn = self.variant(schedule.piecewise_interpolate_schedule(
+      'linear', 200., {5: 1.5, 10: 0.25}))
```
Comment (mtthss) [inline optax/_src/schedule_test.py:310]: Wrong hanging indentation: add two spaces

File: optax/_src/schedule_test.py [hunk 4]
```diff
@@ -302,5 +302,90 @@ def test_decay_count_greater_count_with_alpha(self):
...
+    generated_vals = [schedule_fn(step) for step in range(13)]
+    expected_vals = [200., 220., 240., 260., 280., 300., 255., 210., 165.,
+                     120., 75.,  75.,  75.]
+    np.testing.assert_allclose(generated_vals, expected_vals, atol=1e-3)
+
+  @chex.all_variants()
+  def test_cos_piecewise(self):
+    schedule_fn = self.variant(schedule.piecewise_interpolate_schedule(
+      'cosine', 400., {5: 1.2, 3: 0.6, 7: 1.}))
```
Comment (mtthss) [inline optax/_src/schedule_test.py:319]: Wrong hanging indentation: add two spaces

File: optax/_src/schedule_test.py [hunk 5]
```diff
@@ -302,5 +302,90 @@ def test_decay_count_greater_count_with_alpha(self):
...
+    generated_vals = [schedule_fn(step) for step in range(9)]
+    expected_vals = [400., 360., 280., 240., 264., 288., 288., 288., 288.]
+    np.testing.assert_allclose(generated_vals, expected_vals, atol=1e-3)
+
+  @chex.all_variants()
+  def test_empty_dict(self):
+    schedule_fn = self.variant(schedule.piecewise_interpolate_schedule(
+      'linear', 13., {}))
```
Comment (mtthss) [inline optax/_src/schedule_test.py:327]: Wrong hanging indentation: add two spaces

File: optax/_src/schedule_test.py [hunk 6]
```diff
@@ -302,5 +302,90 @@ def test_decay_count_greater_count_with_alpha(self):
...
+    generated_vals = [schedule_fn(step) for step in range(5)]
+    expected_vals = [13., 13., 13., 13., 13.]
+    np.testing.assert_allclose(generated_vals, expected_vals, atol=1e-3)
+
+  @chex.all_variants()
+  def test_no_dict(self):
+    schedule_fn = self.variant(schedule.piecewise_interpolate_schedule(
+      'cosine', 17.))
```
Comment (mtthss) [inline optax/_src/schedule_test.py:335]: Wrong hanging indentation: add two spaces

File: optax/_src/schedule_test.py [hunk 7]
```diff
@@ -302,5 +302,90 @@ def test_decay_count_greater_count_with_alpha(self):
...
 
+class PiecewiseInterpolateTest(chex.TestCase):
+
+  @chex.all_variants()
+  def test_linear_piecewise(self):
+    schedule_fn = self.variant(schedule.piecewise_interpolate_schedule(
+      'linear', 200., {5: 1.5, 10: 0.25}))
+    generated_vals = [schedule_fn(step) for step in range(13)]
+    expected_vals = [200., 220., 240., 260., 280., 300., 255., 210., 165.,
+                     120., 75.,  75.,  75.]
```
Comment (mtthss) [inline optax/_src/schedule_test.py:313]: Only one space required after comma

**Comment by mtthss** [comment]
If you could fix a couple of small formatting issues I can then merge the PR!
Thanks a lot for contributing these!

File: optax/_src/schedule.py (hunk 1 above)
Comment (n2cholas) [reply optax/_src/schedule.py:263]: Done!

File: optax/_src/schedule.py (hunk 2 above)
Comment (n2cholas) [reply optax/_src/schedule.py]: Done!

File: optax/_src/schedule_test.py (hunk 3 above)
Comment (n2cholas) [reply optax/_src/schedule_test.py]: Done!

File: optax/_src/schedule_test.py (hunk 4 above)
Comment (n2cholas) [reply optax/_src/schedule_test.py]: Done!

**Comment by n2cholas** [comment]
Thanks @mtthss for the review! I've fixed the formatting and resolved all your comments.

**Review by mtthss — APPROVED** [review]

**Comment by mtthss** [comment]
This is now merged! Thanks for your contribution!
